src/opm/parser/eclipse/EclipseState/InitConfig/FoamConfig.cpp

```diff
@@ -17,62 +17,110 @@
-        : reference_surfactant_concentration_( record.getItem( 0 ).getSIDouble( 0 ) )
-        , exponent_( record.getItem( 1 ).getSIDouble( 0 ) )
-        , minimum_surfactant_concentration_( record.getItem( 2 ).getSIDouble( 0 ) )
-    {}
-
-    double FoamRecord::referenceSurfactantConcentration() const {
-        return this->reference_surfactant_concentration_;
-    }
-
-    double FoamRecord::exponent() const {
-        return this->exponent_;
-    }
-
-    double FoamRecord::minimumSurfactantConcentration() const {
-        return this->minimum_surfactant_concentration_;
+namespace Opm
+{
+
+// FoamData member functions.
+
+FoamData::FoamData(const DeckRecord& FOAMFSC_record, const DeckRecord& FOAMROCK_record)
+    : reference_surfactant_concentration_(FOAMFSC_record.getItem(0).getSIDouble(0))
+    , exponent_(FOAMFSC_record.getItem(1).getSIDouble(0))
+    , minimum_surfactant_concentration_(FOAMFSC_record.getItem(2).getSIDouble(0))
+    , allow_desorption_(static_cast<FoamAllowDesorption>(FOAMROCK_record.getItem(0).get<int>(0)))
+    , rock_density_(FOAMROCK_record.getItem(1).getSIDouble(0))
+{
+    // Check validity of adsorption index.
+    const int ads_ind = FOAMROCK_record.getItem(0).get<int>(0);
+    if (ads_ind < 1 || ads_ind > 2) {
+        throw std::runtime_error("Illegal adsorption index in FOAMROCK, must be 1 or 2.");
     }
-
-    // FoamConfig member functions.
-
-    FoamConfig::FoamConfig( const Deck& deck )
-    {
-        if (deck.hasKeyword<ParserKeywords::FOAMFSC>()) {
-            const auto& kw = deck.getKeyword<ParserKeywords::FOAMFSC>();
-            this->records = std::vector<FoamRecord>(kw.begin(), kw.end());
+}
+
+double
+FoamData::referenceSurfactantConcentration() const
+{
+    return this->reference_surfactant_concentration_;
+}
+
+double
+FoamData::exponent() const
+{
+    return this->exponent_;
+}
+
+double
+FoamData::minimumSurfactantConcentration() const
+{
+    return this->minimum_surfactant_concentration_;
+}
+
+FoamData::FoamAllowDesorption
+FoamData::allowDesorption() const
+{
+    return this->allow_desorption_;
+}
+
+double
+FoamData::rockDensity() const
+{
+    return this->rock_density_;
+}
+
+// FoamConfig member functions.
+
+FoamConfig::FoamConfig(const Deck& deck)
+{
+    if (deck.hasKeyword<ParserKeywords::FOAMFSC>()) {
+        const auto& kw_foamfsc = deck.getKeyword<ParserKeywords::FOAMFSC>();
+        if (!deck.hasKeyword<ParserKeywords::FOAMROCK>()) {
+            throw std::runtime_error("FOAMFSC present but no FOAMROCK keyword found.");
+        }
+        const auto& kw_foamrock = deck.getKeyword<ParserKeywords::FOAMROCK>();
+        if (kw_foamfsc.size() != kw_foamrock.size()) {
+            throw std::runtime_error("FOAMFSC and FOAMROCK keywords have different number of records.");
+        }
+        const int num_records = kw_foamfsc.size();
+        this->data_.reserve(num_records);
+        for (int record = 0; record < num_records; ++record) {
```

Comment:
I think the index variable should be called something else - this code is virtually full of `record` variables which are `DeckRecord` instances - so this is slightly misleading.

The `DeckKeyword` class has `begin()` and `end()` methods - but when you need to iterate over two keywords the use of an integer (or maybe `std::size_t` ??) is probably the best.

Reply:
Using `record_index` now.

While I prefer range-for, I do not see a nice way of iterating over both keywords simulateously without the traditional for-loop here.

Reply:
>  I do not see a nice way of iterating over both keywords simulateously without the traditional for-loop here.

I agree - I just wanted to make sure you were aware of the `::begin()` and `::end()` methods of the `DeckKeyword`